pkg/cmos/cmos.go

```diff
@@ -16,18 +16,32 @@ const (
 	cmosDataPort = 0x71
 )
 
+type Cmos struct {
```

Comment:
This should probably be an interface. I am not sure why it was not before. Mocking would be a lot easier.

I'm kinda puzzled at this whole package, actually. 

possible ways to redo it and make it more idiomatic.

Following the Go pattern of file io (and esp. since some CMOS but not be writable!)

type CMOSReader interface {
In(...) error
}

type CMOSWriter interface {
Out (..) error
}

type CMOSReaderWriter {
CMOSReader
CMOSWriter
}

 (go rule: "interfaces should be narrow") -- this is very unlike C++ classes, where the class contains an entire garbage dump of virtual functions -- in Go, you build narrow interfaces and compose them as needed to form more complex interfaces. The anonymous members make this easy.

Your concrete type is then
type CMOSChip struct {
f *os.File
address io.SectionReader
data SectionReader
}
 Take a look at SectionReaders -- even if they may not be applicable they're worth knowing about -- but you might be able to open /dev/mem and then create address and data section readers. Or you might have a different type that implements the CMOS interface: 
type CMOSPort struct {
address uint16
data uint16
}

And then on x86 you can do an iopl system call and the do inb and outb to do IO.

Then you have
func NewCMOSChip(devicepath string, other stuff uint16) (*CMOSChip, error) {
open stuff, create struct
return stuff
}

The new hotness seems to be to return the concrete type, not the interface type. It used to be the other way around, see the go source.

Anyway, how do you ensure that your CMOSChip type implements CMOS? with a type assertion
var _ CMOS = &CMOSChip{}

weird but true. 


Your mocking type is then
type CMOSMock struct {
fakeitup string
}

and your assertion
var _ = &CMOSMock{}

Note that CMOS is CMOS, not Cmos, --> CMOS is an acronym -- Complementary Metal Oxide Semiconductor -- so CMOS.

Anyway, see what you think of these ideas and let's iterate.

Reply:
Yeah, but what you may forgot or overlooked is the fact, that the Reader and Writer interface has to be part of the memio package, not CMOS. The interface, if so, should MemIOReader, MemIOWriter, but this is not scope of this package.

I will mark this package and MemIO as: Worth to rework.

Reply:
oh I no doubt forgot things :-), thanks.